README.rst
Python rfc5424 syslog logging handler
=====================================

.. image:: https://readthedocs.org/projects/rfc5424-logging-handler/badge/?version=latest
    :target: https://rfc5424-logging-handler.readthedocs.io/en/latest/?badge=latest
    :alt: Documentation Status
.. image:: https://img.shields.io/pypi/v/rfc5424-logging-handler.svg
    :target: https://pypi.python.org/pypi/rfc5424-logging-handler
.. image:: https://img.shields.io/pypi/pyversions/rfc5424-logging-handler.svg
    :target: https://pypi.python.org/pypi/rfc5424-logging-handler#downloads
.. image:: https://travis-ci.org/jobec/rfc5424-logging-handler.svg?branch=master
    :target: https://travis-ci.org/jobec/rfc5424-logging-handler
.. image:: https://codecov.io/github/jobec/rfc5424-logging-handler/coverage.svg?branch=master
    :target: https://codecov.io/github/jobec/rfc5424-logging-handler?branch=master

An up-to-date, `RFC 5424 <https://tools.ietf.org/html/rfc5424>`_ compliant syslog handler for the Python logging framework.

* Free software: BSD License
* Homepage: https://github.com/jobec/rfc5424-logging-handler
* Documentation: http://rfc5424-logging-handler.readthedocs.org/

Features
--------

* `RFC 5424 <https://tools.ietf.org/html/rfc5424>`_ Compliant.
* Python Logging adapter for easier sending of rfc5424 specific fields.
* No need for complicated formatting strings.
* TLS/SSL syslog support.
* Alternate transports like streams (ex. stderr, stdout, file, ...).

Installation
------------

Python package::

    pip install rfc5424-logging-handler

Usage
-----

After installing you can use this package like this:

.. code-block:: python

    import logging
    from rfc5424logging import Rfc5424SysLogHandler

    logger = logging.getLogger('syslogtest')
    logger.setLevel(logging.INFO)

    sh = Rfc5424SysLogHandler(address=('10.0.0.1', 514))
    logger.addHandler(sh)

    logger.info('This is an interesting message', extra={'msgid': 'some_unique_msgid'})

This will send the following message to the syslog server::

    <14>1 2020-01-01T05:10:20.841485+01:00 myserver syslogtest 5252 some_unique_msgid - \xef\xbb\xbfThis is an interesting message

Note the UTF8 Byte order mark (BOM) preceding the message. While required by
`RFC 5424 section 6.4 <https://tools.ietf.org/html/rfc5424#section-6.4>`_ if the message is known to be UTF-8 encoded,
there are still syslog receivers that cannot handle it. To bypass this limitation, when initializing the handler Class,
set the ``msg_as_utf8`` parameter to ``False`` like this:

.. code-block:: python

    sh = Rfc5424SysLogHandler(address=('10.0.0.1', 514), msg_as_utf8=False)

For more examples, have a look at the `documentation <http://rfc5424-logging-handler.readthedocs.org/>`_


CHANGELOG.rst
Changelog
---------

`1.4.3`_ - 2019/05/19
~~~~~~~~~~~~~~~~~~~~~

**Changed**

* `#34`_ Allow enterprise ID to contain sub-identifiers.

**Fixed**

* `#31`_ Correct handling of failed connection attempt in TCP transport handler.

`1.4.2`_ - 2019/04/08
~~~~~~~~~~~~~~~~~~~~~

**Changed**

* `#32`_ ``address`` can now also be a list, making loading settings form a config file possible.

**Added**

* Python 3.7 tests and support.

`1.4.1`_ - 2019/02/16
~~~~~~~~~~~~~~~~~~~~~

**Fixed**

* `#29`_ Fix AttributeError when using TLS connection.

`1.4.0`_ - 2019/01/30
~~~~~~~~~~~~~~~~~~~~~

**Added**

* `#27`_ Make it possible to log to streams as an alternate transport.
* Added API documentation.

**Changed**

* Syslog facilities and framing options have moved from the ``RfcSysLogHandler`` class
  to module level variables. **You may have to adjust your references to them.**

`1.3.0`_ - 2018/10/19
~~~~~~~~~~~~~~~~~~~~~

**Added**

* `#23`_ Add support for TLS/SSL

`1.2.1`_ - 2018/09/21
~~~~~~~~~~~~~~~~~~~~~

**Fixed**

* `#21`_ Registered structured data IDs where also suffixed with an enterprise ID.

**Added**

* `#22`_ Add ``utc_timestamp`` parameter to allow logging in UTC time.

`1.1.2`_ - 2018/02/03
~~~~~~~~~~~~~~~~~~~~~

**Fixed**

* `#15`_ When logging to ``/dev/log`` with python 2.7, the connection was permanently lost when the local syslog server
  was restarted.
* `#16`_ The ``extra`` info of a message did not overwrite that of the logging adapter instance.

`1.1.1`_ - 2017/12/08
~~~~~~~~~~~~~~~~~~~~~

**Fixed**

* `#14`_ Fixed handling of ``extra`` parameter in logging adapter.

`1.1.0`_ - 2017/11/24
~~~~~~~~~~~~~~~~~~~~~

**Added**

* The ``msg`` parameter for the logger handler can now be absent allowing "structured data only" messages.

**Fixed**

* Correct the automatic value of the ``hostname`` when the value is anything other then ``NILVALUE``
* The syslog message is now empty in conformance with RFC5424 when it's value is ``None`` or an empty string.


`1.0.3`_ - 2017/10/08
~~~~~~~~~~~~~~~~~~~~~

No functional changes. Only documentation was changed.

**Added**

* Logstash configuration example for RFC5424.

**Changed**

* Moved most of the documentation out of the readme file.

`1.0.2`_ - 2017/08/31
~~~~~~~~~~~~~~~~~~~~~

**Fixed**

* Package description rendering on PyPi due to bug `pypa/wheel#189 <https://github.com/pypa/wheel/issues/189>`_

`1.0.1`_ - 2017/08/30
~~~~~~~~~~~~~~~~~~~~~

**Added**

* `#12`_: It's now possible to send syslog messages as `MSG-ANY <https://tools.ietf.org/html/rfc5424#section-6>`_
  which suppresses the UTF-8 byte order mark (BOM) when sending messages.

`1.0.0`_ - 2017/05/30
~~~~~~~~~~~~~~~~~~~~~

**Changed**

* `#10`_: Procid, appname and hostname can now be set per message, both with the handler as well as with the adapter

.. note::
   This release has a slight change in behaviour. Setting one of the appnama, hostname of procid message to None or an
   empty string will cause it to be filled in automatically. Previously, setting it to an empty string caused it to
   be set to NILVALUE (a - ). You now need to set it explicilty to NILVALUE if you want to omit it from the message.

`0.2.0`_ - 2017/01/27
~~~~~~~~~~~~~~~~~~~~~

**Fixed**

* Better input handling
* Better sanitizing of invalid input

`0.1.0`_ - 2017/01/22
~~~~~~~~~~~~~~~~~~~~~

**Added**

* `#4`_: Adapter class to make it easier to log message IDs or structured data
* Logging of EMERGENCY, ALERT and NOTICE syslog levels by using the adapter class
* Extensive test suite

`0.0.2`_ - 2017/01/18
~~~~~~~~~~~~~~~~~~~~~

**Added**

* `#5`_ Introduced Python 2.7 compatibility

0.0.1 - 2017/01/11
~~~~~~~~~~~~~~~~~~

* Initial release

.. _1.4.3: https://github.com/jobec/rfc5424-logging-handler/compare/1.4.2...1.4.3
.. _1.4.2: https://github.com/jobec/rfc5424-logging-handler/compare/1.4.1...1.4.2
.. _1.4.1: https://github.com/jobec/rfc5424-logging-handler/compare/1.4.0...1.4.1
.. _1.4.0: https://github.com/jobec/rfc5424-logging-handler/compare/1.3.0...1.4.0
.. _1.3.0: https://github.com/jobec/rfc5424-logging-handler/compare/1.2.1...1.3.0
.. _1.2.1: https://github.com/jobec/rfc5424-logging-handler/compare/1.1.2...1.2.1
.. _1.1.2: https://github.com/jobec/rfc5424-logging-handler/compare/1.1.1...1.1.2
.. _1.1.1: https://github.com/jobec/rfc5424-logging-handler/compare/1.1.0...1.1.1
.. _1.1.0: https://github.com/jobec/rfc5424-logging-handler/compare/1.0.3...1.1.0
.. _1.0.3: https://github.com/jobec/rfc5424-logging-handler/compare/1.0.2...1.0.3
.. _1.0.2: https://github.com/jobec/rfc5424-logging-handler/compare/1.0.1...1.0.2
.. _1.0.1: https://github.com/jobec/rfc5424-logging-handler/compare/1.0.0...1.0.1
.. _1.0.0: https://github.com/jobec/rfc5424-logging-handler/compare/0.2.0...1.0.0
.. _0.2.0: https://github.com/jobec/rfc5424-logging-handler/compare/0.1.0...0.2.0
.. _0.1.0: https://github.com/jobec/rfc5424-logging-handler/compare/0.0.2...0.1.0
.. _0.0.2: https://github.com/jobec/rfc5424-logging-handler/compare/0.0.1...0.0.2


rfc5424logging/handler.py
# coding=utf-8
import socket
import sys
from codecs import BOM_UTF8
from collections import OrderedDict
from datetime import datetime
from logging import Handler

from pytz import utc
from tzlocal import get_localzone

from rfc5424logging import transport

NILVALUE = '-'

SP = b' '
# As defined in RFC5424 Section 7
REGISTERED_SD_IDs = ('timeQuality', 'origin', 'meta')
SYSLOG_VERSION = '1'

EMERGENCY = 70
EMERG = EMERGENCY
ALERT = 60
NOTICE = 25

PY2 = sys.version_info[0] == 2


#  facility codes
LOG_KERN = 0       # kernel messages
LOG_USER = 1       # random user-level messages
LOG_MAIL = 2       # mail system
LOG_DAEMON = 3     # system daemons
LOG_AUTH = 4       # security/authorization messages
LOG_SYSLOG = 5     # messages generated internally by syslogd
LOG_LPR = 6        # line printer subsystem
LOG_NEWS = 7       # network news subsystem
LOG_UUCP = 8       # UUCP subsystem
LOG_CRON = 9       # clock daemon
LOG_AUTHPRIV = 10  # security/authorization messages (private)
LOG_FTP = 11       # FTP daemon
#  other codes through 15 reserved for system use
LOG_LOCAL0 = 16    # reserved for local use
LOG_LOCAL1 = 17    # reserved for local use
LOG_LOCAL2 = 18    # reserved for local use
LOG_LOCAL3 = 19    # reserved for local use
LOG_LOCAL4 = 20    # reserved for local use
LOG_LOCAL5 = 21    # reserved for local use
LOG_LOCAL6 = 22    # reserved for local use
LOG_LOCAL7 = 23    # reserved for local use


class Rfc5424SysLogHandler(Handler):
    """
    A handler class which sends RFC 5424 formatted logging records to a syslog server.

    Based on the python built-in SyslogHandler class, but simplified in some parts.
    """
    # from <linux/sys/syslog.h>:
    # ======================================================================
    # priorities/facilities are encoded into a single 32-bit quantity, where
    # the bottom 3 bits are the priority (0-7) and the top 28 bits are the
    # facility (0-big number). Both the priorities and the facilities map
    # roughly one-to-one to strings in the syslogd(8) source code.  This
    # mapping is included in this file.
    #
    # priorities (these are ordered)
    LOG_EMERG = 0    # system is unusable
    LOG_ALERT = 1    # action must be taken immediately
    LOG_CRIT = 2     # critical conditions
    LOG_ERR = 3      # error conditions
    LOG_WARNING = 4  # warning conditions
    LOG_NOTICE = 5   # normal but significant condition
    LOG_INFO = 6     # informational
    LOG_DEBUG = 7    # debug-level messages

    priority_map = {
        "DEBUG": LOG_DEBUG,
        "INFO": LOG_INFO,
        "NOTICE": LOG_NOTICE,
        "WARNING": LOG_WARNING,
        "ERROR": LOG_ERR,
        "CRITICAL": LOG_CRIT,
        "ALERT": LOG_ALERT,
        "EMERGENCY": LOG_EMERG,
        "EMERG": LOG_EMERG,
    }

    def __init__(
            self,
            address=('localhost', transport.SYSLOG_PORT),
            facility=LOG_USER,
            socktype=socket.SOCK_DGRAM,
            framing=transport.FRAMING_NON_TRANSPARENT,
            msg_as_utf8=True,
            hostname=None,
            appname=None,
            procid=None,
            structured_data=None,
            enterprise_id=None,
            utc_timestamp=False,
            timeout=5,
            tls_enable=False,
            tls_ca_bundle=None,
            tls_verify=True,
            tls_client_cert=None,
            tls_client_key=None,
            tls_key_password=None,
            stream=None,
    ):
        """
        Returns a new instance of the Rfc5424SysLogHandler class intended to communicate with
        a remote machine whose address is given by address in the form of a (host, port) tuple.
        If address is not specified, ``('localhost', 514)`` is used. The address is used to open a
        socket.

        An alternative to providing a (host, port) tuple is providing an address as a
        string, for example `/dev/log`. In this case, a Unix domain socket is used to send the
        message to the syslog. If facility is not specified, LOG_USER is used. The type of
        socket opened depends on the socktype argument, which defaults to socket.SOCK_DGRAM
        and thus opens a UDP socket. To open a TCP socket (for use with the newer syslog
        daemons such as rsyslog), specify a value of socket.SOCK_STREAM.

        Note that if your server is not listening on UDP port 514, SysLogHandler may appear
        not to work. In that case, check what address you should be using for a domain socket
        - it's system dependent. For example, on Linux it's usually ``/dev/log`` but on OS/X
        it's ``/var/run/syslog``. You'll need to check your platform and use the appropriate
        address (you may need to do this check at runtime if your application needs to run
        on several platforms). On Windows, you pretty much have to use the UDP option.

        As an alternative transport, you can also provide a stream

        Args:
            address (tuple|list):
                address in the form of a (host, port) tuple or list

            facility (int):
                One of the ``rfc5424logging.LOG_*`` values.

            socktype (int):
                One of ``socket.SOCK_STREAM`` (TCP) or ``socket.SOCK_DGRAM`` (UDP).

            framing (int, optional):
                One of the ``rfc5424logging.FRAMING_*`` values according to
                RFC6587 section 3.4. Only applies when sockettype is ``socket.SOCK_STREAM`` (TCP)
                and is used to give the syslog server an indication about the boundaries
                of the message. Defaults to ``FRAMING_NON_TRANSPARENT`` which will escape all
                newline characters in the message and end the message with a newline character.
                When set to ``FRAMING_OCTET_COUNTING``, it will prepend the message length to the
                begin of the message.
            msg_as_utf8 (bool):
                Controls the way the message is sent.
                disabling this parameter sends the message as MSG-ANY (RFC2424 section 6), avoiding
                issues with receivers that don't support the UTF-8 Byte Order Mark (BOM) at
                the beginning of the message.
            hostname (str):
                The hostname of the system where the message originated from.
                Defaults to the values returned by ``socket.gethostname()``
            appname (str):
                The name of the application. Defaults to the name of the logger that sent
                the message.
            procid (any):
                The process ID of the sending application. Defaults to the ``process`` attribute
                of the log record.
            structured_data (dict):
                A dictionary with structured data that is added to every message. Per message your
                can add more structured data by adding it to the ``extra`` argument of the log function.
            enterprise_id (str):
                The Private Enterprise Number. This is used to compose the structured data IDs when
                they do not include an Enterprise ID and are not one of the reserved structured data IDs.
                Can be a single PEN like ``32473`` or optionally contain sub-identifiers like ``32473.2.6``
            utc_timestamp (bool):
                Whether the timestamp should be converted to UTC time or kept in the local timezone
            timeout (int):
                Sets the timeout on the connection to the server.
            tls_enable (bool):
                If set to ``True``, it sets up a TLS/SSL connection to the address specified in ``address``
                over which the syslog messages will be sent.
                Default to ``False``
            tls_ca_bundle (str):
                The path to a bundle of CA certificates used for validating the remote server's identity.
                If set to ``None``, it will try to load the default CA as described in
                https://docs.python.org/3/library/ssl.html#ssl.SSLContext.load_verify_locations
            tls_verify (bool):
                Whether to verify the certificate of the server.
            tls_client_cert (str):
                path to a file containing a client certificate.
            tls_client_key (str):
                Path to a file containing the client private key.
            tls_key_password (str):
                Optionally the password for decrypting the specified private key.
            stream (io.BufferedIOBase, file, io.TextIOBase):
                Optionally a stream object to send the message to. See https://docs.python.org/3/library/io.html
                for details.
        """
        super(Rfc5424SysLogHandler, self).__init__()

        self.address = address
        self.facility = facility
        self.socktype = socktype
        self.socket = None
        self.unixsocket = False
        self.hostname = hostname
        self.appname = appname
        self.procid = procid
        self.structured_data = structured_data
        self.enterprise_id = enterprise_id
        self.framing = framing
        self.msg_as_utf8 = msg_as_utf8
        self.utc_timestamp = utc_timestamp
        self.timeout = timeout
        self.tls_enable = tls_enable
        self.tls_ca_bundle = tls_ca_bundle
        self.tls_verify = tls_verify
        self.tls_client_cert = tls_client_cert
        self.tls_client_key = tls_client_key
        self.tls_key_password = tls_key_password
        self.stream = stream
        self.transport = None

        if not (isinstance(self.facility, int) and LOG_KERN <= self.facility <= LOG_LOCAL7):
            raise ValueError("Facility is not valid")

        if self.hostname is None or self.hostname == '':
            self.hostname = socket.gethostname()
        if not isinstance(self.structured_data, dict):
            self.structured_data = OrderedDict()

        self._setup_transport()

    def _setup_transport(self):
        if self.stream is not None:
            self.transport = transport.StreamTransport(self.stream)
        elif isinstance(self.address, str):
            self.transport = transport.UnixSocketTransport(self.address, self.socktype)
        elif isinstance(self.address, (tuple, list)):
            if self.socktype == socket.SOCK_STREAM:
                if self.tls_enable:
                    self.transport = transport.TLSSocketTransport(
                        self.address, self.timeout, self.framing,
                        self.tls_ca_bundle, self.tls_verify,
                        self.tls_client_cert, self.tls_client_key, self.tls_key_password
                    )
                else:
                    self.transport = transport.TCPSocketTransport(self.address, self.timeout, self.framing)
            else:
                self.transport = transport.UDPSocketTransport(self.address, self.timeout)
        else:
            raise ValueError("Unsupported address type")

    def encode_priority(self, facility, priority):
        """
        Encode the facility and priority. You can pass in strings or
        integers - if strings are passed, the facility_names and
        priority_names mapping dictionaries are used to convert them to
        integers.
        """
        return (facility << 3) | self.priority_map.get(priority, self.LOG_WARNING)

    @staticmethod
    def filter_printusascii(str_to_filter):
        return ''.join([x for x in str_to_filter if 33 <= ord(x) <= 126])

    def get_hostname(self, record):
        hostname = getattr(record, 'hostname', None)
        if not hostname:
            hostname = self.hostname
        if hostname is None or hostname == '':
            hostname = NILVALUE
        return self.filter_printusascii(str(hostname))

    def get_appname(self, record):
        appname = getattr(record, 'appname', self.appname)
        if appname is None or appname == '':
            appname = getattr(record, 'name', NILVALUE)
        return self.filter_printusascii(str(appname))

    def get_procid(self, record):
        procid = getattr(record, 'procid', self.procid)
        if procid is None or procid == '':
            procid = getattr(record, 'process', NILVALUE)
        return self.filter_printusascii(str(procid))

    def get_msgid(self, record):
        msgid = getattr(record, 'msgid', NILVALUE)
        if msgid is None or msgid == '':
            msgid = NILVALUE
        return self.filter_printusascii(str(msgid))

    def get_enterprise_id(self, record):
        # We allow None to be returned here.
        # We'll handle it when cleaning the structured data
        enterprise_id = getattr(record, 'enterprise_id', self.enterprise_id)
        if enterprise_id is None:
            return None
        else:
            return self.filter_printusascii(str(enterprise_id))

    def get_structured_data(self, record):
        structured_data = OrderedDict()
        if isinstance(self.structured_data, dict):
            structured_data.update(self.structured_data)
        record_sd = getattr(record, 'structured_data', {})
        if isinstance(record_sd, dict):
            structured_data.update(record_sd)
        return structured_data

    def build_msg(self, record):
        # The syslog message has the following ABNF [RFC5234] definition:
        #
        #     SYSLOG-MSG      = HEADER SP STRUCTURED-DATA [SP MSG]
        #
        #     HEADER          = PRI VERSION SP TIMESTAMP SP HOSTNAME
        #                     SP APP-NAME SP PROCID SP MSGID
        #     PRI             = "<" PRIVAL ">"
        #     PRIVAL          = 1*3DIGIT ; range 0 .. 191
        #     VERSION         = NONZERO-DIGIT 0*2DIGIT
        #     HOSTNAME        = NILVALUE / 1*255PRINTUSASCII
        #
        #     APP-NAME        = NILVALUE / 1*48PRINTUSASCII
        #     PROCID          = NILVALUE / 1*128PRINTUSASCII
        #     MSGID           = NILVALUE / 1*32PRINTUSASCII
        #
        #     TIMESTAMP       = NILVALUE / FULL-DATE "T" FULL-TIME
        #     FULL-DATE       = DATE-FULLYEAR "-" DATE-MONTH "-" DATE-MDAY
        #     DATE-FULLYEAR   = 4DIGIT
        #     DATE-MONTH      = 2DIGIT  ; 01-12
        #     DATE-MDAY       = 2DIGIT  ; 01-28, 01-29, 01-30, 01-31 based on
        #                             ; month/year
        #     FULL-TIME       = PARTIAL-TIME TIME-OFFSET
        #     PARTIAL-TIME    = TIME-HOUR ":" TIME-MINUTE ":" TIME-SECOND
        #                     [TIME-SECFRAC]
        #     TIME-HOUR       = 2DIGIT  ; 00-23
        #     TIME-MINUTE     = 2DIGIT  ; 00-59
        #     TIME-SECOND     = 2DIGIT  ; 00-59
        #     TIME-SECFRAC    = "." 1*6DIGIT
        #     TIME-OFFSET     = "Z" / TIME-NUMOFFSET
        #     TIME-NUMOFFSET  = ("+" / "-") TIME-HOUR ":" TIME-MINUTE
        #
        #
        #     STRUCTURED-DATA = NILVALUE / 1*SD-ELEMENT
        #     SD-ELEMENT      = "[" SD-ID *(SP SD-PARAM) "]"
        #     SD-PARAM        = PARAM-NAME "=" %d34 PARAM-VALUE %d34
        #     SD-ID           = SD-NAME
        #     PARAM-NAME      = SD-NAME
        #     PARAM-VALUE     = UTF-8-STRING ; characters '"', '\' and
        #                                  ; ']' MUST be escaped.
        #     SD-NAME         = 1*32PRINTUSASCII
        #                     ; except '=', SP, ']', %d34 (")
        #
        #     MSG             = MSG-ANY / MSG-UTF8
        #     MSG-ANY         = *OCTET ; not starting with BOM
        #     MSG-UTF8        = BOM UTF-8-STRING
        #     BOM             = %xEF.BB.BF
        #
        #     UTF - 8 - STRING = *OCTET ; UTF - 8 string as specified
        #                               ; in RFC 3629
        #
        #     OCTET = % d00 - 255
        #     SP = % d32
        #     PRINTUSASCII = % d33 - 126
        #     NONZERO - DIGIT = % d49 - 57
        #     DIGIT = % d48 / NONZERO - DIGIT
        #     NILVALUE = "-"

        # HEADER
        pri = '<%d>' % self.encode_priority(self.facility, record.levelname)
        version = SYSLOG_VERSION
        timestamp = datetime.fromtimestamp(record.created, get_localzone())
        if self.utc_timestamp:
            timestamp = timestamp.astimezone(utc)
        timestamp = timestamp.isoformat()
        hostname = self.get_hostname(record)
        appname = self.get_appname(record)
        procid = self.get_procid(record)
        msgid = self.get_msgid(record)

        pri = pri.encode('ascii')
        version = version.encode('ascii')
        timestamp = timestamp.encode('ascii')
        hostname = hostname.encode('ascii', 'replace')[:255]
        appname = appname.encode('ascii', 'replace')[:48]
        procid = procid.encode('ascii', 'replace')[:128]
        msgid = msgid.encode('ascii', 'replace')[:32]

        header = b''.join((pri, version, SP, timestamp, SP, hostname, SP, appname, SP, procid, SP, msgid))

        # STRUCTURED-DATA
        enterprise_id = self.get_enterprise_id(record)
        structured_data = self.get_structured_data(record)
        cleaned_structured_data = []
        for sd_id, sd_params in list(structured_data.items()):
            # Clean structured data ID
            sd_id = self.filter_printusascii(sd_id)
            sd_id = sd_id.replace('=', '').replace(' ', '').replace(']', '').replace('"', '')
            if '@' not in sd_id and sd_id not in REGISTERED_SD_IDs and enterprise_id is None:
                raise ValueError("Enterprise ID has not been set. Cannot build structured data ID. "
                                 "Please set a enterprise ID when initializing the logging handler "
                                 "or include one in the structured data ID.")
            elif '@' in sd_id:
                sd_id, enterprise_id = sd_id.rsplit('@', 1)

            if len(enterprise_id) > 30:
                raise ValueError("Enterprise ID is too long. Impossible to build structured data ID.")

            sd_id = sd_id.replace('@', '')
            if len(sd_id) + len(enterprise_id) > 32:
                sd_id = sd_id[:31 - len(enterprise_id)]
            if sd_id not in REGISTERED_SD_IDs:
                sd_id = '@'.join((sd_id, enterprise_id))
            sd_id = sd_id.encode('ascii', 'replace')

            cleaned_sd_params = []
            # ignore sd params not int key-value format
            if isinstance(sd_params, dict):
                sd_params = sd_params.items()
            else:
                sd_params = []

            # Clean key-value pairs
            for (param_name, param_value) in sd_params:
                param_name = self.filter_printusascii(str(param_name))
                param_name = param_name.replace('=', '').replace(' ', '').replace(']', '').replace('"', '')
                if param_value is None:
                    param_value = ''

                param_value = str(param_value)

                if PY2:
                    param_value = unicode(param_value, 'utf-8')  # noqa

                param_value = param_value.replace('\\', '\\\\').replace('"', '\\"').replace(']', '\\]')

                param_name = param_name.encode('ascii', 'replace')[:32]
                param_value = param_value.encode('utf-8', 'replace')

                sd_param = b''.join((param_name, b'="', param_value, b'"'))
                cleaned_sd_params.append(sd_param)

            cleaned_sd_params = SP.join(cleaned_sd_params)

            # build structured data element
            spacer = SP if cleaned_sd_params else b''
            sd_element = b''.join((b'[', sd_id, spacer, cleaned_sd_params, b']'))
            cleaned_structured_data.append(sd_element)

        if cleaned_structured_data:
            structured_data = b''.join(cleaned_structured_data)
        else:
            structured_data = NILVALUE.encode('ascii')

        # MSG
        if record.msg is None or record.msg == "":
            pieces = (header, structured_data)
        else:
            msg = self.format(record)
            if self.msg_as_utf8:
                msg = b''.join((BOM_UTF8, msg.encode('utf-8')))
            else:
                msg = msg.encode('utf-8')
            pieces = (header, structured_data, msg)
        syslog_msg = SP.join(pieces)

        return syslog_msg

    def emit(self, record):
        """
        Emit a record.

        The record is formatted, and then sent to the syslog server. If
        exception information is present, it is NOT sent to the server.
        """
        try:
            syslog_msg = self.build_msg(record)
            self.transport.transmit(syslog_msg)
        except Exception:
            self.handleError(record)

    def close(self):
        """
        Closes the socket.
        """
        self.acquire()
        try:
            if self.transport is not None:
                self.transport.close()
            super(Rfc5424SysLogHandler, self).close()
        finally:
            self.release()
